### ISSUE TITLE
Deprecation of AsyncAPI Generator CLI - removing `ag`, not the library 

### ISSUE THREAD

**Issue by derberg** [asyncapi/generator#1008]
tl;dr

- we have [AsyncAPI CLI](https://github.com/asyncapi/cli) and `asyncapi generate fromTemplate` 
- all missing features are added to AsyncAPI CLI, just [docs need update](https://github.com/asyncapi/generator/issues?q=is%3Aopen+label%3A%22good+first+issue%22+label%3A%22%F0%9F%93%91+docs%22)
- we no longer invest dev time in `ag`

So what I'm suggesting is basically:
- remove `ag`, so `cli.js` and watch related code
- create an issue in every template repository for people to followup and remove any instruction about `ag`

What I do not know:
- what is the best way to do it.... like it is not a standalone package that we can mark as deprecated in `npm` as `@asyncapi/generator` is basically both now, library and cli
- maybe we should join it with my proposed rename https://github.com/asyncapi/generator/issues/580 so we deprecate `@asyncapi/generator` on npm and publish the library alone as `@asyncapi/generator-core`. This might be the best for all the users that use AsyncAPI Generator CLI in CI/CD pipelines. My experience tells me that many people, when use CLIs in CI/CD, do not care much about versioning and install latest. So for example if we remove cli and release it as `2.0`, still under `@asyncapi/generator` then many CI/CD pipelines around the world will explode 😄 which is not nice 😄

**Comment by fmvilas** [comment]
What about leaving the `ag` CLI but it always returns an error code and prints a message saying it's deprecated and you should use AsyncAPI CLI? Then in some time from now we just completely remove it.

**Comment by derberg** [comment]
@fmvilas the problem is that we will break many pipelines, this is why even releasing major will be harmful.

> My experience tells me that many people, when use CLIs in CI/CD, do not care much about versioning and install latest. So for example if we remove cli and release it as 2.0, still under @asyncapi/generator then many CI/CD pipelines around the world will explode 😄 which is not nice 😄

Of course we can always say that it is not our fault that people use the tool wrong, but that ain't cool. If we know we might cause issues, we should try to avoid it, thus I think the best would be to simply release `@asyncapi/generator-core` and deprecate entire `@asyncapi/generator`.

**Comment by smoya** [comment]
What about importing the CLI in this generator, and call the generate logic through it? We could ass a warning message saying this package is deprecated and keep the generator alive for a while. CLI dependence will be updated automatically, meaning feature parity. 
We could keep it alive until a new major release of the CLI happens if that works.

**Comment by derberg** [comment]
@smoya keep in mind that CLI goal is to work just as a CLI and not be used as library.

> and call the generate logic through it

this means `exec` the binary

**Comment by smoya** [comment]
> @smoya keep in mind that CLI goal is to work just as a CLI and not be used as library.

But why adding such a constraint? Is there any strong reason of limiting that? Isn't possible to extract that logic into a proper service that can be called from code?

**Comment by smoya** [comment]
In the very worst case, https://oclif.io/docs/running_programmatically

**Comment by fmvilas** [comment]
> I think the best would be to simply release @asyncapi/generator-core and deprecate entire @asyncapi/generator.

Yeah, I vote for this too then.

**Comment by smoya** [comment]
> > I think the best would be to simply release @asyncapi/generator-core and deprecate entire @asyncapi/generator.
> 
> Yeah, I vote for this too then.

So just for my understanding, will then that library be used in CLI?

**Comment by derberg** [comment]
> But why adding such a constraint? Is there any strong reason of limiting that? Isn't possible to extract that logic into a proper service that can be called from code? I see only benefits, but for sure there is a reason for not allowing it that I'm missing.

if there will be a strong use case that will be worth the effort, sure, why not. But definitely not worth doing with only assumption it makes sense.

> So just for my understanding, will then that library be used in CLI?

yes, it will, it has to. We would just change name from `@asyncapi/generator` to `@asyncapi/generator-core` and the new version would not contain the CLI

**Comment by smoya** [comment]
> yes, it will, it has to. We would just change name from `@asyncapi/generator` to `@asyncapi/generator-core` and the new version would not contain the CLI

Makes totally sense then. I like the approach 🙌

**Comment by anandsunderraman** [comment]
@derberg What is the process to migrate a generator template from `ag` to `cli` ?

**Comment by lmgyuan** [comment]
Does this issue still need help? Based on my understanding of it, it looks like there might still be work needed to update the readme and documentation to reflect the changes and guide users on how to transition from ag to the AsyncAPI CLI. Is my understanding correct? If so, I am happy to help! : )
